pkg/gadgets/trace/exec/tracer/tracer_test.go

```diff
@@ -0,0 +1,214 @@
+
+	tests := map[string]utilstest.Test[types.Event]{
+		"test without filtering": {
+			GetTracer: func(info *utilstest.Info, eventCallback func(event types.Event)) utilstest.Tracer {
+				tracer, err := tracer.NewTracer(&tracer.Config{}, nil, eventCallback)
+				if err != nil {
+					testRunner.Fatalf("Error creating tracer: %s", err)
+				}
+				return tracer
+			},
+			GenerateEvent: generateEvent,
+			GetAnyExpectedEvent: func(info *utilstest.Info) *types.Event {
+				return &types.Event{
+					Event: eventtypes.Event{
+						Type: eventtypes.NORMAL,
+					},
+					Pid:       uint32(catPid),
+					Ppid:      uint32(info.Pid),
+					UID:       uint32(info.UID),
+					MountNsID: info.MountNsID,
+					Retval:    0,
+					Comm:      "cat",
+					Args:      []string{"/bin/cat", "/dev/null"},
+				}
+			},
+		},
+		"test filtering (no event should be captured)": {
+			GetTracer: func(info *utilstest.Info, eventCallback func(event types.Event)) utilstest.Tracer {
+				m := utilstest.CreateMntNsFilterMap(t, 0)
+				tracer, err := tracer.NewTracer(&tracer.Config{MountnsMap: m}, nil, eventCallback)
+				if err != nil {
+					testRunner.Fatalf("Error creating tracer: %s", err)
+				}
+				return tracer
+			},
+			GenerateEvent: generateEvent,
+		},
+		"test filtering": {
+			GetTracer: func(info *utilstest.Info, eventCallback func(event types.Event)) utilstest.Tracer {
+				m := utilstest.CreateMntNsFilterMap(t, info.MountNsID)
+				tracer, err := tracer.NewTracer(&tracer.Config{MountnsMap: m}, nil, eventCallback)
+				if err != nil {
+					testRunner.Fatalf("Error creating tracer: %s", err)
+				}
+				return tracer
+			},
+			GenerateEvent: generateEvent,
+			GetExpectedEvent: func(info *utilstest.Info) *types.Event {
+				return &types.Event{
+					Event: eventtypes.Event{
+						Type: eventtypes.NORMAL,
+					},
+					Pid:       uint32(catPid),
+					Ppid:      uint32(info.Pid),
+					UID:       uint32(info.UID),
+					MountNsID: info.MountNsID,
+					Retval:    0,
+					Comm:      "cat",
+					Args:      []string{"/bin/cat", "/dev/null"},
+				}
+			},
+		},
+		"test UID": {
+			GetTracer: func(info *utilstest.Info, eventCallback func(event types.Event)) utilstest.Tracer {
+				m := utilstest.CreateMntNsFilterMap(t, info.MountNsID)
+				tracer, err := tracer.NewTracer(&tracer.Config{MountnsMap: m}, nil, eventCallback)
+				if err != nil {
+					testRunner.Fatalf("Error creating tracer: %s", err)
+				}
+				return tracer
+			},
+			RunnerConfig:  &utilstest.Config{UID: 1435},
+			GenerateEvent: generateEvent,
+			GetExpectedEvent: func(info *utilstest.Info) *types.Event {
+				return &types.Event{
+					Event: eventtypes.Event{
+						Type: eventtypes.NORMAL,
+					},
+					Pid:       uint32(catPid),
+					Ppid:      uint32(info.Pid),
+					UID:       uint32(info.UID),
+					MountNsID: info.MountNsID,
+					Retval:    0,
+					Comm:      "cat",
+					Args:      []string{"/bin/cat", "/dev/null"},
+				}
+			},
+		},
+		"test many args": {
```

Comment:
Does this test validates some different code path from the tests above?

Maybe it should be named something like: `Tracer captures at most 19 command arguments`? Then more arguments should be provided for the command, but less should be expected.

Or, what happens if there is more than 19 arguments? Will the tracing fail? I guess describing whether it fails or is being truncated is a very important detail here from the user perspective.

The reason I'm asking is in general, that single unit test should consist of single assertion (excluding setup code), targeting single behavior in the code base (certain condition or certain side effect etc.).

Reply:
You're totally right. I changed the test to call the binary with more than 19 arguments but check that only 19 are captured.